File: Top Interview 150/C++/Majority_Element.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        short maxOccurences = 0;
        int majorityElement = 0;
        vector<int> uniqueNums = unique(nums);
        for(int i = 0; i < uniqueNums.size(); i++){
            int occurences = countOccurences(nums, uniqueNums[i]);
            if(occurences > maxOccurences){
                maxOccurences = occurences;
                majorityElement = uniqueNums[i];
            }
        }
        return majorityElement;
    }
private:
    int countOccurences(vector<int>& nums, int num){
        int count = 0;
        for(int i = 0; i < nums.size(); i++){
            if(nums[i] == num){
                count++;
            }
        }
        return count;
    }
    vector<int> unique(vector<int>& nums){
        vector<int> unique;
        for(int i = 0; i < nums.size(); i++){
            if(find(unique.begin(), unique.end(), nums[i]) == unique.end()){
                unique.push_back(nums[i]);
            }
        }
        return unique;
    }
};

int main(){
    Solution solution;
    vector<int> nums = {3,2,3};
    cout << solution.majorityElement(nums) << endl; // 3
    return 0;
}

File: Top Interview 150/Python/Majority_Element.py
class Solution:
    def majorityElement(self, nums: list[int]) -> int:
        unique_nums = set(nums)
        max = 0
        max_num = 0
        for num in unique_nums:
            if nums.count(num) > max:
                max = nums.count(num)
                max_num = num
        return max_num

print(Solution().majorityElement([3,2,3])) # 3
print(Solution().majorityElement([2,2,1,1,1,2,2])) # 2
print(Solution().majorityElement([1])) # 1
print(Solution().majorityElement([6,5,5])) # 5